docs/docs/reconciliation.md

```diff
@@ -0,0 +1,123 @@
+---
+id: reconciliation
+title: Reconciliation
+permalink: docs/reconciliation.html
+---
+
+React provides a declarative API so that you don't have to worry about exactly what changes on every update. This makes writing applications a lot easier, but it might not be obvious how this is implemented within React. This article explains the choices we made in React's "diffing" algorithm so that component updates are predictable while being fast enough for high-performance apps.
+
+## Motivation
+
+When you use React, at a single point in time you can think of the `render()` function as creating a tree of React elements. On the next state or props update, that `render()` function will return a different tree of React elements. The React framework then needs to figure out a way to efficiently convert the first tree into the second tree.
+
+There are some generic solutions to this algorithmic problem of generating the minimum number of operations to transform one tree into another. However, the [state of the art algorithms](http://grfia.dlsi.ua.es/ml/algorithms/references/editsurvey_bille.pdf) have a complexity in the order of O(n<sup>3</sup>) where n is the number of elements in the tree.
+
+If we used this in React, displaying 1000 elements would require in the order of one billion comparisons. This is far too expensive. Instead, React implements a heuristic O(n) algorithm based on two assumptions:
+
+1. Two elements of different types can be treated as totally different trees.
+2. The developer can hint at which child elements may be stable across different renders with a `key` prop.
+
+In practice, these assumptions are valid for almost all practical use cases.
+
+## The Diffing Algorithm
+
+When diffing two trees, React first compares the types of the two root elements. If the root elements have different types, React just throws away the first tree and builds the second tree from scratch. When the root elements have the same type, React first converts the root element, and then recurses. The specifics are different depending on whether the elements are DOM elements or component elements.
+
+### DOM Elements
+
+When comparing two React DOM elements, React looks at the attributes of both, keeps the same underlying DOM element, and only updates the changed attributes. For example:
```

Comment:
Let's add a section in the beginning that React will destroy the old DOM node and create a new one if their types don't match. For example, `<span>` is replaced with a `<button>`. Any components below also get unmounted, and their state is destroyed. For example, replacing `<div><Counter /></div>` with `<span><Counter /></span>` will reset `Counter`'s state and re-mount it.

After that, we explain what happens if type is the same.